### ISSUE TITLE
Distinguish `Shape` and `Bounds` in `Locatable`

### ISSUE THREAD

**Issue by averbraeck** [averbraeck/dsol#33]
Bounds for quick intersection (visibility in viewport), Shape for click and interaction.

**Comment by averbraeck** [comment]
After multiple discussions, The following points became clear:

`Location` (a `Point<?>`) is a property of the **simulation object**. Only the simulation object knows where it is, how it is moving, and possibly what its direction vector is.

`Bounds` are in essence a property of the animation object. Only it knows how it is being drawn, and what the extent of is drawing is. Unfortunately, when bounds would be linked to the animation, they are not a property in world coordinates but in pixel coordinates. They would also lack a z-extension (a height) in a 2d-animation, which would result in not being able to draw objects that overlap with a 'height slice', e.g., the first floor of a building. Therefore, the `Bounds` remain linked to the **simulation object**.

Anything that has to do with clicking on an object or interacting with an object using the medium on which the object is presented, takes place in the 'rendered' world, in other words, in the **animation object** which is in our case a `Renderable`.

**Comment by averbraeck** [comment]
A `Renderable` object for a screen could in theory implement `getClickableShape()` or `getClickableBounds()` For a circular object, this is, however, not ideal. There, the distance to the center of the circle is the only thing we need, and not a 64-segment polygon resembling the circle with a `contains()` method.

A fast fail-method `contains(x, y)` for `Renderable`, where `x` and `y` are in unscaled screen coordinates, would be ideal to implement the functionality to test which objects have been clicked on a mouse-down or mouse-up event. The programmer can then also decide to allow a few pixels 'around' the object to still be clickable, or to return `false` when the zoom factor has resulted in an object to be too small to be seen on the screen.

**Comment by averbraeck** [comment]
Currently, the `Renderable2dInterface` has the method `boolean contains(Point2d pointWorldCoordinates, Bounds2d extent);`.  This is, however, a very difficult method to implement. It would better be replaced by a backcasted method using screen coordinates for the unscaled, non-translated and non-rotated drawn object.

**Comment by averbraeck** [comment]
So, the method that would be needed is translating back the mouse click position relative to the extent (viewport) to the unscaled object. This can be done using the following steps:
1. Obtain the mouse click in screen coordinates, relative to the `Dimension` of the viewport. Let's call this coordinate _(sx, sy)_.
2. Transform the `Location` _(wx, wy)_ in world coordinates of the source object to screen coordinates. For this, the method `RenderableScale.getScreenCoordinates(final Point<?> worldCoordinates, final Bounds2d extent, final Dimension screen)` can be used. Let's call this coordinate _(ox, oy)_.
3. Translate the point _(sx, sy)_ to a relative coordinate _(sx', sy')_ using _sx' = sx - ox_, _sy' = sy - oy_. Now the point _(sx', sy')_ is relative to the screen location of the location point of the object on the screen.
4. Rotate back the relative location of the point _(sx', sy')_ to its drawing orientation, by rotating it with an angle `-source.getRotZ()`. The new location is _(sx", sy")_
5. Scale back the relative unrotated location of the point _(sx", sy")_ to its also unscaled position, by calculating: _mx = sx" / scaleX_, _my = sy" / scaleY_. The point (mx, my) is the screen coordinate of the mouse, relative to the originally drawn object in the `graphics2D.paint()` method, and in pixel coordinates. 
6. Offer the point _(mx, my)_ to an abstract method `Renderable.contains(mx, my)` that will determine of the point _(mx, my)_ is close enough to the drawn object (contained, on the edge, within a boundary) of the drawn object on the screen.
7. A default implementation of `Renderable.contains(mx, my)` compares the point _(mx, my)_ with the projected bounds of the source object around (0, 0).

**Comment by averbraeck** [comment]
For clicking on the object with a margin, there is a difference between applying the margin on the screen coordinates or on the object coordinates. When it is applied on the screen coordinates, there is a margin of, say, m pixels around the object that translate to (z * m) pixels on the object coordinates, where z is the zoom factor. One could also take the object coordinates as the basis, e.g., having a margin of 10% around the object. When the object is zoomed in on the screen, this results in more 'pixels' around the object than when the object is zoomed out. Both methods can have their advantages and disadvantages:
- when there are many very small objects on the screen, the screen-based method results in a lot of overlapping margins, and hence, multiple objects that appear to be clicked when clearly one of them is clicked. Clicking work well with the object-based method.
- when there are few very small objects on the screen, the screen-based method works well, but the object-based method makes it hard to click on the object.
- when an object is zoomed in massively, the object-based method results in a margin that is too large, whereas the screen-based method works fine.

In other words, it is situation dependent, and can therefore best be left to the modeler. What this analysis makes clear is that the method for clicking on the object needs to know the zoom factor.

**Comment by averbraeck** [comment]
The **current method** to test for a click is the `contains` method. The `contains` method makes the calculation in world coordinates (so, the mouse point click has already been transformed from screen coordinates to world coordinates). 

```java
public boolean contains(final Point2d pointWorldCoordinates, final Bounds2d extent)
```

It currently calculates the results purely based on the bounds of the object (`source.getBounds()`):

```java
Bounds2d intersect = BoundsUtil.projectBounds(this.source.getLocation(), this.source.getBounds());
return intersect.contains(pointWorldCoordinates);
```

The method `BoundsUtil.projectBounds` transforms a bounds (4 corner points) from the object coordinates relative to (0,0) to a bounds in world coordinates, using the rotation of the object (`source.getDirZ()`) and its translation (`source.getLocation()`) and calculates a new bounds in absolute world coordinates based on this calculation. When rotating the object, the bounds become larger as a result.

**Comment by averbraeck** [comment]
The `contains(screen)` method is now documented as follows:

```java
    /**
     * This method determines whether the shape drawn on the screen (rotated, zoomed, translated) 
     * contain the provided point that is provided in screen coordinates (e.g., the position of a mouse 
     * click on the screen).
     * @param pointScreenCoordinates Point2D; the point in screen coordinates, e.g. from a mouse event; 
     * note that the point is given in AWT/Swing coordinates, so as a `Point2D` and not a `Point2d` 
     * @param extent Bounds2d; the extent of the viewport that is visible on the screen, in world coordinates
     * @param screenSize Dimension; the dimensions of the screen, in screen coordinates
     * @param scale RenderableScale; the current zoom factor of the screen, as well as the y-scale ratio
     * @param margin double; the margin to apply 'around' the object, in screen coordinates
     * @param relativeMargin boolean; whether the margin should be scaled with the zoom factor (true) 
     * or not (false)
     * @return whether the point is in the shape or in a margin around the shape
     */
```

**Comment by averbraeck** [comment]
The `contains(screen)` method transforms the object to (0,0) with a direction of 0 radians and scale of 1, and also transforms the (clicked) point to this coordinate system. After this, it calls a delegate method:

```java
return contains(pointRelativeTo00, scale, worldMargin, pixelMargin, xScale, yScale);
```

This is the method to override for alternative, non-rectangular 'contains' implementations. Note that this implementation has to take into account a lot of different elements when determining whether a pixel is 'inside' or 'outside' the region of interest:
- the `worldMargin`, i.e., the margin in unscaled world coordinates (e.g., 5 meters around the object).
- the `pixelMargin`, i.e., the margin in pixels around the object, independent of the zoom factor on the screen, and also independent of the y-scale factor.
- the object scale factor (`scale.getObjectScaleFactor()`). This scale factor determines whether the object is drawn 'smaller' or 'larger' on the screen than its actual implementation in the `paint()` method.
- `xScale`, the translation of a pixel to a world coordinate in the x-direction, given the zoom factor and y-scale factor.
- `yScale`, the translation of a pixel to a world coordinate in the y-direction, given the zoom factor and y-scale factor.

**Comment by averbraeck** [comment]
The reference implementation of the above method is:

```java
/**
 * Reference implementation of the contains method that uses the bounding box to determine 
 * whether the shape contains the point (e.g., a mouse click) or not.
 * @param pointRelativeTo00 Point2d; the point relative to the drawing world.
 * @param scale RenderableScale; the current zoom factor of the screen
 * @param worldMargin double; the margin to apply 'around' the object, in screen coordinates 
 *     at a zoom level of 1, which is the same as world coordinates. This margin grows and shrinks 
 *     in absolute sense with the zoom factor.
 * @param pixelMargin double; the number of pixels around the drawn object for contains 
 *     to be 'true'. This guarantees that a mouse click can be pointed to a very small object.
 * @param xScale double; the ratio between a world x-coordinate and a pixel
 * @param yScale double; the ratio between a world y-coordinate and a pixel
 * @return whether the point is in the shape or in a margin around the shape
 */
public boolean contains(final Point2d pointRelativeTo00, final RenderableScale scale, 
        final double worldMargin, final double pixelMargin, final double xScale, final double yScale)
{
    try
    {
        Bounds<?, ?> b = getSource().getBounds();
        Bounds2d bounds = new Bounds2d(b.getMinX() * scale.getObjectScaleFactor(), 
                b.getMaxX() * scale.getObjectScaleFactor(), b.getMinY() * scale.getObjectScaleFactor(), 
                b.getMaxY() * scale.getObjectScaleFactor());
        double xMarginWorld = Math.max(worldMargin * scale.getObjectScaleFactor(), pixelMargin * xScale);
        double yMarginWorld = Math.max(worldMargin * scale.getObjectScaleFactor(), pixelMargin * yScale);
        Bounds2d marginBounds = new Bounds2d(bounds.getMinX() - xMarginWorld, 
                bounds.getMaxX() + xMarginWorld, bounds.getMinY() - yMarginWorld, 
                bounds.getMaxY() + yMarginWorld);
        return marginBounds.covers(pointRelativeTo00);
    }
    catch (RemoteException exception)
    {
        CategoryLogger.always().warn(exception, "contains");
        return false;
    }
}
```

**Comment by averbraeck** [comment]
Note that the `contains` methods have been tested with **object scaling** and with an **y-scale factor**, e.g., as follows to give the y-scale a compression of 2.0, and to make the objects half the size:

```java
getAnimationTab().getAnimationPanel().setRenderableScale(new RenderableScale(2.0, 0.5));
```

To show how the new contains method can be used, see the below examples:

In the animation object, the following override takes care that clicking 5 'units' (e.g., meters) around the object is recognized as a 'contains'. This scales with the object, so it is 5 'units' in world coordinates:

```java
    @Override
    public boolean contains(final Point2D pointScreenCoordinates, final Bounds2d extent, final Dimension screenSize,
            final RenderableScale scale, final double worldMargin, final double pixelMargin)
    {
        return super.contains(pointScreenCoordinates, extent, screenSize, scale, 5.0, 0.0);
    }
```

In the animation object, the following override takes care that clicking 50 pixels around the object, independent of the zoom level, is recognized as a 'contains'. 

```java
    @Override
    public boolean contains(final Point2D pointScreenCoordinates, final Bounds2d extent, final Dimension screenSize,
            final RenderableScale scale, final double worldMargin, final double pixelMargin)
    {
        return super.contains(pointScreenCoordinates, extent, screenSize, scale, 0.0, 50.0);
    }
```

Note that both can be combined. The method uses the maximum number of pixels from the two, given the zoom level. Suppose we want to click 2 units (e.g., meters) around the object, but when the object becomes really small, we still want 4 pixels in all directions round the bounding box:

```java
    @Override
    public boolean contains(final Point2D pointScreenCoordinates, final Bounds2d extent, final Dimension screenSize,
            final RenderableScale scale, final double worldMargin, final double pixelMargin)
    {
        return super.contains(pointScreenCoordinates, extent, screenSize, scale, 2.0, 4.0);
    }
```

**Comment by averbraeck** [comment]
The contains method seems to work according to the design.
